### ISSUE TITLE
Why do we need an extra classification head

### ISSUE THREAD

**Issue by dongdongtong** [Wenlve-Zhou/VLP-UDA#1]
Great work! Thanks for open source this impressive work. I have one question about the knowledge distillation.

As we know, in the original arch of CLIP, two branches exits (one text branch, another is vision branch). When finetuning CLIP in UDA, a straightforward way is to fix the text branch and tuning the vision branch using source/target data. In this tuning process, the original contrastive InfoNCE loss should be used to align the vision features to the text features. We train the vision branch because we want the vision branch to extract domain-invariant features, and we fix the text branch is for the reason that the natural language, i.e., the text features, are already domain-invariant. However, in this work, you add an extra task head on the top of the vision branch and train it to be capable of classifying different images. I am wondering the motivation why do you not train the two branch directly and train the extra task head instead? Is this two-branch training leading to relatively bad results?

Best regards.

**Comment by Wenlve-Zhou** [comment]
Hello, thank you very much for your recognition of our work. In fact, we designed it this way to ensure the scalability of our approach. You may notice that the form of CMKD is very simple, and when combined with other methods, such as Fixmatch used in our paper, it can further improve performance. In UDA methods based on pure visual pre-training models, retaining the Backbone and adding new task heads is the most common paradigm. Therefore, we retained this concept, with the purpose of making it easy for subsequent researchers to combine some existing methods with ours. Secondly, considering UDA techniques based on VLP models, methods like DAPL will use prompt tuning to fine-tune the CLIP model. You can find that in the current paradigm of CMKD, it can also be easily combined with such methods. Therefore, in general, we are not committed to creating an optimal method, but rather hope to design a highly scalable Baseline for subsequent researchers to study. I hope this can answer your question. Best wishes!



---- Replied Message ----
| From | ***@***.***> |
| Date | 09/14/2024 20:17 |
| To | ***@***.***> |
| Cc | ***@***.***> |
| Subject | [Wenlve-Zhou/VLP-UDA] Why do we need an extra classification head (Issue #1) |

Great work! Thanks for open source this impressive work. I have one question about the knowledge distillation.

As we know, in the original arch of CLIP, two branches exits (one text branch, another is vision branch). When finetuning CLIP in UDA, a straightforward way is to fix the text branch and tuning the vision branch using source/target data. In this tuning process, the original contrastive InfoNCE loss should be used to align the vision features to the text features. We train the vision branch because we want the vision branch to extract domain-invariant features, and we fix the text branch is for the reason that the natural language, i.e., the text features, are already domain-invariant. However, in this work, you add an extra task head on the top of the vision branch and train it to be capable of classifying different images. I want to know the motivation why you do not train the two branch directly and train the task head instead? Is this two-branch training leading to bad results?

Best regards.

**Comment by dongdongtong** [comment]
Thanks for your patient answer :) .

I have previously read your paper, so I definitely understand your design for scalability :) .

Unfortunately, I also make some tests based on the res50 backbone of CLIP on the office-home dataset. I have frozen the batch norm layers of res50 (set required_grad to false, set batch norm to .eval when training). I used the context learning (CoOp, a well-known prompt tuning method tailored for CLIP) for the text branch, turned on all other parameters of res50 except BN, and used original contrastive learning objective. 

However, this full parameter training recipe for the vision branch leads to worse results. For example, in art to clip_art exp, only ~50 acc was obtained. But I found that your paper reports a high performance using the extra task head with help of the text branch. So I am wondering that the performance gap between the extra task head and this two-branch training :) . I may make some mistakes. So just for kind consult, did you have ever tried this two-branch training recipe and obtain a high performance?

best regards~

**Comment by Wenlve-Zhou** [comment]
I believe I understand your point now. You're saying that you've used our method as a baseline and trained CLIP with COOP, but the results were not satisfactory, correct?


If that's the case, I'd like to offer you some advice:


1. We did experiment with incorporating the text model into the training, and we maintained a good results, which is not much different from the results presented in the paper. However, please note that we performed Fine-tuning on the text encoder without using COOP. Therefore, it is feasible to add the text encoder to the training without causing the model to collapse. If you have reproduced our results, I suggest you first try Fine-tuning the text encoder to verify the stability of the algorithm.


2. Regarding the adjustment of the text encoder, I have some suggestions for you. If you choose to Fine-tune, the learning rate for the text encoder should also be controlled to be relatively low. Secondly, although I have not tried combining COOP with cmkd, in some of my new work this year, I have been thinking about prompt tuning or learnable token, and I hope to offer some insights beyond this paper. You should pay attention to the initialization of your learnable token; normal initialization is often more stable than truncated normal initialization. Additionally, be mindful if you have added weight decay to the learnable token; adding weight decay might lead to poor results.


3. If you have not used our method and simply want to ask if I can Fine-tune CLIP's text encoder and visual encoder, then I think you need a better way to tune parameters. Because the UDA dataset is too small, it is easy to lose a lot of pre-trained knowledge. You can try using a smaller learning rate for the visual encoder, or you can use the RST we proposed to train the visual encoder. Since RST can greatly preserve pre-trained parameters, it does not lose pre-trained knowledge.
I can only offer you so much advice, as more suggestions might involve work that I have not yet published. There is a piece of work that might be published soon, also about how to more effectively utilize the pre-trained part. You can keep an eye on our latest news on GitHub after it is accepted. I hope these suggestions are helpful to you!






---- Replied Message ----
| From | ***@***.***> |
| Date | 09/16/2024 11:28 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [Wenlve-Zhou/VLP-UDA] Why do we need an extra classification head (Issue #1) |

Thanks for your patient answer :) .

I have previously read your paper, so I definitely understand your design for scalability :) .

Unfortunately, I also make some tests based on the res50 backbone of CLIP on the office-home dataset. I have frozen the batch norm layers of res50 (set required_grad to false, set batch norm to .eval when training). I used the context learning (CoOp, a well-known prompt tuning method tailored for CLIP) for the text branch, turned on all other parameters of res50 except BN, and used original contrastive learning objective.

However, this full parameter training recipe for the vision branch leads to worse results. For example, in art to clip_art exp, only ~50 acc was obtained. But I found that your paper reports a high performance using the extra task head with help of the text branch. So I am wondering that the performance gap between the extra task head and this two-branch training :) . I may make some mistakes. So just for kind consult, did you have ever tried this two-branch training recipe and obtain a high performance?

**Comment by dongdongtong** [comment]
Many thanks for your suggestions, I am actually standing at the advice 3. I fix the text branch and have tried different learning rates ranges from 3e-7 to 3e-3. So I doulbt the problem may exist in the text branch. I believe we have a happy communication. Wishes for your new works!
